File: test/cpu/cmp2.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"
#include "test.hpp"

#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;

BOOST_FIXTURE_TEST_SUITE(CMP2, Prepare)
struct F {
    F() {
        // CMP2.B (%A2), %D3
        TEST::SET_W(0, 0000320 | 2);
        TEST::SET_W(2, 3 << 12);
        TEST::SET_W(4, TEST_BREAK);

        // CMP2.B (%A2), %A3
        TEST::SET_W(6, 0000320 | 2);
        TEST::SET_W(8, 1 << 15 | 3 << 12);
        TEST::SET_W(10, TEST_BREAK);

        // CMP2.W (%A2), %D3
        TEST::SET_W(12, 0001320 | 2);
        TEST::SET_W(14, 3 << 12);
        TEST::SET_W(16, TEST_BREAK);

        // CMP2.W (%A2), %A3
        TEST::SET_W(18, 0001320 | 2);
        TEST::SET_W(20, 1 << 15 | 3 << 12);
        TEST::SET_W(22, TEST_BREAK);
        
        // CMP2.L (%A2), %D3
        TEST::SET_W(24, 0002320 | 2);
        TEST::SET_W(26, 3 << 12);
        TEST::SET_W(28, TEST_BREAK);

        // CMP2.L (%A2), %A3
        TEST::SET_W(30, 0002320 | 2);
        TEST::SET_W(32, 1 << 15 | 3 << 12);
        TEST::SET_W(34, TEST_BREAK);

        jit_compile(0, 36);
    }
};
BOOST_AUTO_TEST_SUITE(R, *boost::unit_test::fixture<F>())
BOOST_AUTO_TEST_SUITE(Byte)

BOOST_AUTO_TEST_SUITE(Unsigned)
BOOST_AUTO_TEST_CASE(In) {
    RAM[0x100] = 70;
    RAM[0x101] = 230;
    cpu.A[2] = 0x100;
    cpu.D[3] = 150;
    run_test(0);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Leq) {
    RAM[0x100] = 70;
    RAM[0x101] = 230;
    cpu.A[2] = 0x100;
    cpu.D[3] = 70;
    run_test(0);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Geq) {
    RAM[0x100] = 70;
    RAM[0x101] = 230;
    cpu.A[2] = 0x100;
    cpu.D[3] = 230;
    run_test(0);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}
BOOST_AUTO_TEST_CASE(Under) {
    RAM[0x100] = 70;
    RAM[0x101] = 230;
    cpu.A[2] = 0x100;
    cpu.D[3] = 20;
    run_test(0);
    BOOST_TEST(cpu.C);
}

BOOST_AUTO_TEST_CASE(Over) {
    RAM[0x100] = 70;
    RAM[0x101] = 230;
    cpu.A[2] = 0x100;
    cpu.D[3] = 250;
    run_test(0);
    BOOST_TEST(cpu.C);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Signed)
BOOST_AUTO_TEST_CASE(In) {
    RAM[0x100] = -70;
    RAM[0x101] = 80;
    cpu.A[2] = 0x100;
    cpu.A[3] = 10;
    run_test(6);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Leq) {
    RAM[0x100] = -70;
    RAM[0x101] = 80;
    cpu.A[2] = 0x100;
    cpu.A[3] = -70;
    run_test(6);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Geq) {
    RAM[0x100] = -70;
    RAM[0x101] = 80;
    cpu.A[2] = 0x100;
    cpu.A[3] = 80;
    run_test(6);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}
BOOST_AUTO_TEST_CASE(Under) {
    RAM[0x100] = -70;
    RAM[0x101] = 80;
    cpu.A[2] = 0x100;
    cpu.A[3] = -100;
    run_test(6);
    BOOST_TEST(cpu.C);
}

BOOST_AUTO_TEST_CASE(Over) {
    RAM[0x100] = -70;
    RAM[0x101] = 80;
    cpu.A[2] = 0x100;
    cpu.A[3] = 100;
    run_test(6);
    BOOST_TEST(cpu.C);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE(Word)
BOOST_AUTO_TEST_SUITE(Unsigned)
BOOST_AUTO_TEST_CASE(In) {
    TEST::SET_W(0x100, 7000);
    TEST::SET_W(0x102, 23000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 15000;
    run_test(12);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Leq) {
    TEST::SET_W(0x100, 7000);
    TEST::SET_W(0x102, 23000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 7000;
    run_test(12);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Geq) {
    TEST::SET_W(0x100, 7000);
    TEST::SET_W(0x102, 23000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 23000;
    run_test(12);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}
BOOST_AUTO_TEST_CASE(Under) {
    TEST::SET_W(0x100, 7000);
    TEST::SET_W(0x102, 23000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 2000;
    run_test(12);
    BOOST_TEST(cpu.C);
}

BOOST_AUTO_TEST_CASE(Over) {
    TEST::SET_W(0x100, 7000);
    TEST::SET_W(0x102, 23000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 25000;
    run_test(12);
    BOOST_TEST(cpu.C);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Signed)
BOOST_AUTO_TEST_CASE(In) {
    TEST::SET_W(0x100, -700);
    TEST::SET_W(0x102, 2300);
    cpu.A[2] = 0x100;
    cpu.A[3] = 100;
    run_test(18);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Leq) {
    TEST::SET_W(0x100, -700);
    TEST::SET_W(0x102, 2300);
    cpu.A[2] = 0x100;
    cpu.A[3] = -700;
    run_test(18);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Geq) {
    TEST::SET_W(0x100, -700);
    TEST::SET_W(0x102, 2300);
    cpu.A[2] = 0x100;
    cpu.A[3] = 2300;
    run_test(18);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}
BOOST_AUTO_TEST_CASE(Under) {
    TEST::SET_W(0x100, -700);
    TEST::SET_W(0x102, 2300);
    cpu.A[2] = 0x100;
    cpu.A[3] = -1000;
    run_test(18);
    BOOST_TEST(cpu.C);
}

BOOST_AUTO_TEST_CASE(Over) {
    TEST::SET_W(0x100, -700);
    TEST::SET_W(0x102, 2300);
    cpu.A[2] = 0x100;
    cpu.A[3] = 2600;
    run_test(18);
    BOOST_TEST(cpu.C);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE(Long)
BOOST_AUTO_TEST_SUITE(Unsigned)
BOOST_AUTO_TEST_CASE(In) {
    TEST::SET_L(0x100, 700000);
    TEST::SET_L(0x104, 2300000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 1500000;
    run_test(24);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Leq) {
    TEST::SET_L(0x100, 700000);
    TEST::SET_L(0x104, 2300000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 700000;
    run_test(24);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Geq) {
    TEST::SET_L(0x100, 700000);
    TEST::SET_L(0x104, 2300000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 2300000;
    run_test(24);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}
BOOST_AUTO_TEST_CASE(Under) {
    TEST::SET_L(0x100, 700000);
    TEST::SET_L(0x104, 2300000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 200000;
    run_test(24);
    BOOST_TEST(cpu.C);
}

BOOST_AUTO_TEST_CASE(Over) {
    TEST::SET_L(0x100, 700000);
    TEST::SET_L(0x104, 2300000);
    cpu.A[2] = 0x100;
    cpu.D[3] = 25000000;
    run_test(24);
    BOOST_TEST(cpu.C);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Signed)
BOOST_AUTO_TEST_CASE(In) {
    TEST::SET_L(0x100, -70000);
    TEST::SET_L(0x104, 230000);
    cpu.A[2] = 0x100;
    cpu.A[3] = 10000;
    run_test(30);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Leq) {
    TEST::SET_L(0x100, -70000);
    TEST::SET_L(0x104, 230000);
    cpu.A[2] = 0x100;
    cpu.A[3] = -70000;
    run_test(30);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(Geq) {
    TEST::SET_L(0x100, -70000);
    TEST::SET_L(0x104, 230000);
    cpu.A[2] = 0x100;
    cpu.A[3] = 230000;
    run_test(30);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(!cpu.C);
}
BOOST_AUTO_TEST_CASE(Under) {
    TEST::SET_L(0x100, -70000);
    TEST::SET_L(0x104, 230000);
    cpu.A[2] = 0x100;
    cpu.A[3] = -100000;
    run_test(30);
    BOOST_TEST(cpu.C);
}

BOOST_AUTO_TEST_CASE(Over) {
    TEST::SET_L(0x100, -70000);
    TEST::SET_L(0x104, 230000);
    cpu.A[2] = 0x100;
    cpu.A[3] = 260000;
    run_test(30);
    BOOST_TEST(cpu.C);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/cmp2.lua
-- CMP2.B (%A5), %D1
MEM[0].W = OCT "000320" | 5
MEM[2].W = 1 << 12
MEM[4].W = TEST.BREAK

-- CMP2.B (%A5), %A1
MEM[6].W = OCT "000320" | 5
MEM[8].W = (8 + 1) << 12
MEM[10].W = TEST.BREAK

-- CMP2.W (%A5), %D2
MEM[12].W = OCT "001320" | 5
MEM[14].W = 2 << 12
MEM[16].W = TEST.BREAK

-- CMP2.W (%A5), %A2
MEM[18].W = OCT "001320" | 5
MEM[20].W = (8 + 2) << 12
MEM[22].W = TEST.BREAK

-- CMP2.L (%A5), %D3
MEM[24].W = OCT "002320" | 5
MEM[26].W = 3 << 12
MEM[28].W = TEST.BREAK

-- CMP2.L (%A5), %A3
MEM[30].W = OCT "002320" | 5
MEM[32].W = (8 + 3) << 12
MEM[34].W = TEST.BREAK

JIT_COMPILE(0, 36)

local run = {}
function run.pre() CPU.A[6] = 0x1000 end
run.B = {
    pre = function()
        MEM[0x1000].B = 70
        MEM[0x1001].B = 230
    end,
    inRange = function()
        CPU.D[2].B = 150
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.Z", false)
        TEST.CHECK("CPU.CC.C", false)
    end,
    leq = function()
        CPU.D[2].B = 70
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    geq = function()
        CPU.D[2].B = 230
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    under = function()
        CPU.D[2].B = 20
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.C", true)
    end,
    over = function()
        CPU.D[2].B = 250
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.C", true)
    end
}

run.C = {
    pre = function()
        MEM[0x1000].C = -70
        MEM[0x1001].C = 80
    end,
    inRange = function()
        CPU.A[2] = 10
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.Z", false)
        TEST.CHECK("CPU.CC.C", false)
    end,
    leq = function()
        CPU.A[2] = -70
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    geq = function()
        CPU.A[2] = 80
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    under = function()
        CPU.A[2] = -100
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.C", true)
    end,
    over = function()
        CPU.A[2] = 100
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.C", true)
    end
}

run.W = {
    pre = function()
        MEM[0x1000].W = 7000
        MEM[0x1002].W = 23000
    end,
    inRange = function()
        CPU.D[3].W = 15000
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.Z", false)
        TEST.CHECK("CPU.CC.C", false)
    end,
    leq = function()
        CPU.D[3].W = 7000
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    geq = function()
        CPU.D[3].W = 23000
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    under = function()
        CPU.D[3].W = 2000
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.C", true)
    end,
    over = function()
        CPU.D[3].W = 25000
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.C", true)
    end
}

run.H = {
    pre = function()
        MEM[0x1000].H = -700
        MEM[0x1002].H = 800
    end,
    inRange = function()
        CPU.A[3] = 100
        TEST.RUN(18)
        TEST.CHECK("CPU.CC.Z", false)
        TEST.CHECK("CPU.CC.C", false)
    end,
    leq = function()
        CPU.A[3] = -700
        TEST.RUN(18)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    geq = function()
        CPU.A[3] = 800
        TEST.RUN(18)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    under = function()
        CPU.A[3] = -1000
        TEST.RUN(18)
        TEST.CHECK("CPU.CC.C", true)
    end,
    over = function()
        CPU.A[3] = 1000
        TEST.RUN(18)
        TEST.CHECK("CPU.CC.C", true)
    end
}

run.L = {
    pre = function()
        MEM[0x1000].L = 700000
        MEM[0x1004].L = 2300000
    end,
    inRange = function()
        CPU.D[4].L = 1500000
        TEST.RUN(24)
        TEST.CHECK("CPU.CC.Z", false)
        TEST.CHECK("CPU.CC.C", false)
    end,
    leq = function()
        CPU.D[4].L = 700000
        TEST.RUN(24)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    geq = function()
        CPU.D[4].L = 2300000
        TEST.RUN(24)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    under = function()
        CPU.D[4].L = 200000
        TEST.RUN(24)
        TEST.CHECK("CPU.CC.C", true)
    end,
    over = function()
        CPU.D[4].L = 2500000
        TEST.RUN(24)
        TEST.CHECK("CPU.CC.C", true)
    end
}

run.I = {
    pre = function()
        MEM[0x1000].I = -70000
        MEM[0x1004].I = 80000
    end,
    inRange = function()
        CPU.A[4] = 100
        TEST.RUN(30)
        TEST.CHECK("CPU.CC.Z", false)
        TEST.CHECK("CPU.CC.C", false)
    end,
    leq = function()
        CPU.A[4] = -70000
        TEST.RUN(30)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    geq = function()
        CPU.A[4] = 80000
        TEST.RUN(30)
        TEST.CHECK("CPU.CC.Z", true)
        TEST.CHECK("CPU.CC.C", false)
    end,
    under = function()
        CPU.A[4] = -100000
        TEST.RUN(30)
        TEST.CHECK("CPU.CC.C", true)
    end,
    over = function()
        CPU.A[4] = 100000
        TEST.RUN(30)
        TEST.CHECK("CPU.CC.C", true)
    end
}

return run
